zebra-state/src/memory_state.rs

```diff
@@ -456,3 +457,135 @@ impl Ord for Chain {
         }
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use proptest::collection::vec;
+    use transaction::Transaction;
+
+    use std::mem;
+
+    use zebra_chain::parameters::{Network, NetworkUpgrade};
+    use zebra_chain::serialization::ZcashDeserializeInto;
+    use zebra_test::prelude::*;
+
+    use self::assert_eq;
+    use super::*;
+
+    /// Helper trait for constructing "valid" looking chains of blocks
+    trait FakeChainHelper {
+        fn make_fake_child(&self) -> Arc<Block>;
+    }
+
+    impl FakeChainHelper for Block {
+        fn make_fake_child(&self) -> Arc<Block> {
+            let parent_hash = self.hash();
+            let mut child = Block::clone(self);
+            let mut transactions = mem::take(&mut child.transactions);
+            let mut tx = transactions.remove(0);
+
+            let input = match Arc::make_mut(&mut tx) {
+                Transaction::V1 { inputs, .. } => &mut inputs[0],
+                Transaction::V2 { inputs, .. } => &mut inputs[0],
+                Transaction::V3 { inputs, .. } => &mut inputs[0],
+                Transaction::V4 { inputs, .. } => &mut inputs[0],
+            };
+
+            match input {
+                transparent::Input::Coinbase { height, .. } => height.0 += 1,
+                _ => panic!("block must have a coinbase height to create a child"),
+            }
+
+            child.transactions.push(tx);
+            child.header.previous_block_hash = parent_hash;
+
+            Arc::new(child)
+        }
+    }
+
+    #[test]
+    fn construct_empty() {
+        zebra_test::init();
+        let _chain = Chain::default();
+    }
+
+    #[test]
+    fn construct_single() -> Result<()> {
+        zebra_test::init();
+        let block = zebra_test::vectors::BLOCK_MAINNET_434873_BYTES.zcash_deserialize_into()?;
+
+        let mut chain = Chain::default();
+        chain.push(block);
+
+        assert_eq!(1, chain.blocks.len());
+
+        Ok(())
+    }
+
+    #[test]
+    fn construct_many() -> Result<()> {
+        zebra_test::init();
+
+        let mut block: Arc<Block> =
+            zebra_test::vectors::BLOCK_MAINNET_434873_BYTES.zcash_deserialize_into()?;
```

Comment:
Do we need to use genesis here?
Or does the code we're testing not care about the chain root invariants?

```suggestion
    #[test]
    fn construct_single() -> Result<()> {
        zebra_test::init();
        let block = zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES.zcash_deserialize_into()?;

        let mut chain = Chain::default();
        chain.push(block);

        assert_eq!(1, chain.blocks.len());

        Ok(())
    }

    #[test]
    fn construct_many() -> Result<()> {
        zebra_test::init();

        let mut block: Arc<Block> =
            zebra_test::vectors::BLOCK_MAINNET_GENESIS_BYTES.zcash_deserialize_into()?;
```

Reply:
I don't think we should worry about the genesis block, this is only supposed to ever handle non-finalized data, so the genesis block should never pass through this data structure, or even any pre sapling blocks.